### ISSUE TITLE
Anticheat improvements [discussion]

### ISSUE THREAD

**Issue by Fixer-007** [minetest/minetest#3822]
Current anticheat needs improvement.
What I've seen:
+ slow flying in the air (avoiding "fast" check)
+ infinite blocks 
+ some kind of /give for cheaters
+ using autohealing in combat

Some people seen cheaters bypassing protected doors and chests.
I just want to raise the issue. If you think this is a big issue, feel free to comment.

**Comment by PilzAdam** [comment]
> infinite blocks
> some kind of /give for cheaters

This isn't possible, since the server controls the inventory.

> using autohealing in combat

Health is also server-side. (what happens when the client simply ignores health commands from the server?)

> Some people seen cheaters bypassing protected doors and chests.

Accessing locked chests is not possible.

**Comment by rubenwardy** [comment]
This is needed for CTF.

**Comment by est31** [comment]
One anticheat feature I see that improvement is possible noclip prevention, which can be easily done using check proposed by @Calinou : take the last legal position of the player, take the next one, and check each node on the path whether.

The second anticheat feature I see improvement is possible is no-fly enforcement, it can happen by checking whether there is a non air block in the 18 (9 blocks in x+z, times two for two y layers) blocks below a player... The number of y layers can be increased when the jump height is higher on servers. Of course there needs to be some way to account for people falling (its a real parabole, easiest to see when jumping off high tower with fast priv), they should die, not trigger anticheat :) and for people suggesting this: killing the cheaters is not the solution, as falling people should die at the end of their journey, not in the middle :)
The velocity can't be used as its sent by the clients and untrusted, it can be easily faked. So doing no-fly enforcement is hard, unless sb has a better idea :)

**Comment by ChaosWormz** [comment]
I know minecraft has anticheat mod using bukkit moding api. may we can figure out how that works and implement something similar to that

**Comment by C1ffisme** [comment]
@ChaosWormz Not quite. Minecraft works very, very differently from Minecraft. I imagine this anticheat mod is actually to stop hackers from using hacked clients and cheating in 1000 diamonds.

I think Minecraft has it's own anticheat. I've seen anticheat-like things on servers.

**Comment by 0-afflatus** [comment]
One of the biggest problems with players using hacked clients is the attitude that goes with it. I would like to have some procedural way of flagging undesired behaviour. Enforcement isn't actually quite so important - although preventing noclipping or glitching through protected blocks would be useful. If admin staff could receive a live notification that a player is flying without a license or whatever that would allow them to make more informed decisions.

So often I find myself running on intuition. I banned someone today because I thought they *might* have been using a hacked client, but they may have just been abusing known weaknesses in the engine. If I had had to justify my actions with another admin it would have been useful to be able to say that I *knew* or at least had reasonable evidence that they were using a hacked client.

**Comment by Fixer-007** [comment]
>infinite blocks
>some kind of /give for cheaters
>This isn't possible, since the server controls the inventory.

I've seen cheaters at least on 0.4.12 with huge amounts of gold-bronze-protectors that are impossible to craft by hand in that amount of time.

**Comment by rubenwardy** [comment]
There was a bug that caused entity duplication in 0.4.3, not sure if that was ever fixed

**Comment by ac-minetest** [comment]
I made a noclip anticheat some time ago. Main problems you bump into:

1.with lag server sees you inside stone even if you already digged path to there ( client diggs faster, server is slow to decide whether you are allowed to dig there and all that)
2. you must check pretty regularly or some cheaters will abuse that and slip through
3. if you do 2. you will waste a lot of resources

To solve this i did:
1. remember position when server first sees you inside wall and last safe position  ( not inside wall)
2. ensure that enough time passes to compensate for lag ( do some checks, add to cheat suspicion for that player..). After that do find_path to see if there is path between two points. If no path or too far apart add to cheat suspicion. Once this exceeds certain limit announce cheater.
3. instead of checking all players regularly concentrate of player with most cheat suspicion, check others less frequently. This way very little resources needed.

code:
`local cheat = {};
cheat.players = {};
cheat.suspect = "";
cheat.scan_timer = 0
cheat.deep_scan_timer = 0
cheat.nodelist = {};

local CHEAT_DEEP_TIMESTEP = 2; -- check most suspicious player every 2 seconds
local CHEAT_TIMESTEP = 15; -- check all players less regularly so server burden is minimal
local CHEAT_LIMIT = 5; -- after this player is marked cheater if not found innocent


-- process registered node data to determine walkability of nodes
minetest.after(0,function()
	for name,_ in pairs(minetest.registered_nodes) do
		local def = minetest.registered_nodes[name];
		if def then cheat.nodelist[name] = not(def.walkable) end; -- is true if player can move through node
	end
end)

-- here we check if player can walk inside current position and mark player as cheater if trigger limit exceeded
function check_player(player)
	local pos = player:getpos(); -- feet position
	local name = player:get_player_name();
	local clear = cheat.nodelist[minetest.get_node(pos).name]; -- test for clip violation
	
	--if not clear then minetest.chat_send_all(" not clear for "..name) end
	
	if cheat.players[name].count>CHEAT_LIMIT then -- cheat limit exceeded!
		local dist = (cheat.players[name].clearpos.x-cheat.players[name].cheatpos.x)^2+(cheat.players[name].clearpos.y-cheat.players[name].cheatpos.y)^2+(cheat.players[name].clearpos.z-cheat.players[name].cheatpos.z)^2;
		if dist<256 then -- try to find path between last clear position and first cheat position - there must be path if player really digged through and this is false report cause of lag!
			local path =  minetest.find_path(cheat.players[name].clearpos, pos, 16, 2, 6, "Dijkstra");
			if path then cheat.players[name].count=0 end -- ok, player really could travel between last clear position and problem position
		end
		
		if cheat.players[name].count>0 then
			--round positions for report
			for i,v in pairs(cheat.players[name].clearpos) do cheat.players[name].clearpos[i]=math.ceil(cheat.players[name].clearpos[i]); end
			for i,v in pairs(cheat.players[name].cheatpos) do cheat.players[name].cheatpos[i]=math.ceil(cheat.players[name].cheatpos[i]); end
			local text="Player ".. name .. " was caught walking inside wall between ".. minetest.pos_to_string(cheat.players[name].clearpos) .. " and " .. minetest.pos_to_string(cheat.players[name].cheatpos);
			--minetest.chat_send_all(text);
			minetest.log("action", text);
			-- insert players "reward" here, ideas: loss of interact, death+teleport to spawn+speed reduced to a crawl by override physics..
			cheat.players[name].count=0; -- reset counter
			return;
		end
	
	end
	
	if clear then 
		if cheat.players[name].count>0 then -- very slowly decrease counter
			cheat.players[name].count=cheat.players[name].count-0.1; 
		end
		if cheat.players[name].count == 0 then -- player hasnt "cheated" yet, remember clear position
			cheat.players[name].clearpos = pos 
		end
	else -- player caught inside walls
		local privs = minetest.get_player_privs(name).noclip;--if privs then return end
		local dist = (cheat.players[name].lastpos.x-pos.x)^2+(cheat.players[name].lastpos.y-pos.y)^2+(cheat.players[name].lastpos.z-pos.z)^2;
		if dist>0 then -- if player stands still whole time do nothing
			if cheat.players[name].count == 0 then cheat.players[name].cheatpos = pos end -- remember first position where player found inside wall
			cheat.players[name].count = cheat.players[name].count + 1;
			--minetest.chat_send_all(" counter increased for "..name .. " to ".. cheat.players[name].count)
		end
	end
	cheat.players[name].lastpos = pos
end

minetest.register_globalstep(function(dtime)
	
	cheat.scan_timer = cheat.scan_timer + dtime
	cheat.deep_scan_timer = cheat.deep_scan_timer + dtime
	if cheat.deep_scan_timer < CHEAT_DEEP_TIMESTEP then return end
	cheat.deep_scan_timer = 0
	
	-- MOST SUSPICIOUS PLAYER GETS SPECIAL ATTENTION, AND MORE FREQUENTLY :)
	if cheat.suspect~="" then
		local player = minetest.get_player_by_name(cheat.suspect);
		if player then check_player(player) end
	end
	
	-- GENERAL SCAN OF ALL PLAYERS
	if cheat.scan_timer>CHEAT_TIMESTEP then	
		cheat.scan_timer=0;
		--local t = minetest.get_gametime();
		local players = minetest.get_connected_players();
		local maxcount = 0;	local maxname = "";
		
		for _,player in pairs(players) do
			check_player(player)
			local name = player:get_player_name();
			if cheat.players[name].count>maxcount then -- and the most suspicious player is:
				maxcount = cheat.players[name].count; maxname = name;
			end
		end
		cheat.suspect=maxname;
		--minetest.chat_send_all("suspect is " .. maxname .. " with count " .. maxcount)
	end
end)


minetest.register_on_joinplayer(function(player) -- init stuff on player join
	local name = player:get_player_name(); if name == nil then return end 
	local pos = player:getpos();
	cheat.players[name]={count=0,clearpos = pos,cheatpos = pos, lastpos = pos};
end)`

**Comment by ac-minetest** [comment]
another thing:

if you place blocks quickly and last block is just inside protected area ( with protectod mod for example), it sometimes happens with enough lag that  last block is placed even though "minetest.is_protected = function .... " says it  shouldnt be.. How is that possible?

This was abused recently in just test server to place tnt where it wasnt allowed and wipe out protector blocks.

**Comment by Fixer-007** [comment]
Interesting video with some suspected cheaters: 
https://www.youtube.com/watch?v=8K7nyD-XTGM&feature=youtu.be (look at 30:00 min) for world-edit cheat? This player has only shout and interact privileges, yet ... Note, that was recorded on 0.4.12 with outdated protector mod.

**Comment by JurajVajda** [comment]
i had to disable anti cheat since i wasn't able to use physics override for fast moving - like kind of a potion effect thing :(

**Comment by Fixer-007** [comment]
>There was a bug that caused entity duplication in 0.4.3, not sure if that was ever fixed

Link? I've heared something about duplication, could be related and exploited.

**Comment by orwell96** [comment]
I think we need something like walking-through-walls checking. With most protection mods it is possible to stand on top of a house, dig the node below you, and until the server sends back that this position is protected and the node was not dug you have already fallen through that node and are standing inside the house.
And what about checking is_protected() just when the player starts to dig a node and not when he finishes? So we could avoid nodes being dug in protected areas to disappear for a short time on the client.

**Comment by ShadowNinja** [comment]
I've seen entity duplication bugs in some mods, like pipeworks, which allowed my to get a full chest of mese blocks.  There's nothing the engine can do about this.

**Comment by beyondlimits** [comment]
http://irc.minetest.ru/minetest/2015-05-10#i_4245022

**Comment by est31** [comment]
Note: what I said back then in that discussion is wrong. It is possible to do serverside noclip enforcement, as nofly enforcement. Its not trivial to implement, but possible.

**Comment by Fixer-007** [comment]
This may be useful for server admins:
https://github.com/ac-minetest/anticheat/

**Comment by Calinou** [comment]
> This may be useful for server admins:
https://github.com/ac-minetest/anticheat/

Its full source code is not available though, so not much of an use to Minetest developers.

**Comment by paramat** [comment]
More discussion on this subject in #6114 
Also see this PR #6219

**Comment by Wuzzy2** [comment]
What is the progress of this issue? Is there evidence of the mentioned cheats in recent games?

**Comment by paramat** [comment]
There were some big PRs coming for serverside movement but they were abandoned.
And yes there is still a lot of hacked-client cheating going on.

**Comment by corarona** [comment]
Hey guys,
Hacked Client dev here ( Dragonfire a bit and our own fork: https://repo.or.cz/waspsaliva.git )  i know i'm kinda late to the party and I'm usually not the type to post too much stuff. But I feel like I have to say something.

First off regarding the OP:
 * flying: def possible (just return true on the PrivCheck thingy - same with noclip)
 * infinite blocks/give: There have definitely been dupes in Mineclone (push brewing stand with piston e.g.) but to my knowledge there is no way to have a universal /give thing (as mentioned its serverside) and as far as i know there are no trivial dupes for standard MT (believe me, we've tried ;) )
 * Autohealing: apart from Autoeat (which is def a thing) autohealing in that sense is not possible i dont think – what is possible however is to automate /sethome in a way that it will effectively reduce death penalty to a fraction of a second ( https://repo.or.cz/waspsaliva.git/blob/HEAD:/clientmods/undying/init.lua – a rather crude but functional csm doing it – i know it works on catlandia; i'm not in the habbit of cheating on non Anarchy servers tho so I can't  say how well it works on other servers).

The main reason I'm posting is however, that I feel that you guys don't see that there are legit use cases for hacked clients (i.e. Anarchy Servers) – there is a community for that (even tho it might be small) – and I (and others) have had a lot of fun hacking this thing (and learned a lot too).

From where I'm standing it seems like you (MT devs) should be happy for people doing this (when the source is made available ofc) since you can easily see what cheats are being used right?

Last but not least maybe you guys would have fun playing anarchy too since on the one hand a large part of it is coding and understanding how stuff works for a lot of the players and on the other hand at least the server I'm playing ( clam-ity.minecity.online ) has really cool server culture and a number of players doing really cool things (ingame and hacking wise) – check out our wiki for some Info on the server culture https://clamity.elidragon.com

I would love to open a dialogue about this and please don't hate us so much ^^

-cora

P.S. I have tried to make an account on the MT forums like thrice and was always instantly banned not sure why lol (maybe because i use tor ?)

**Comment by sfan5** [comment]
> what is possible however is to automate /sethome in a way that it will effectively reduce death penalty to a fraction of a second

Yeah, sethome should probably have a cooldown time for itself and also after death.

> From where I'm standing it seems like you (MT devs) should be happy for people doing this (when the source is made available ofc) since you can easily see what cheats are being used right?

Having the source of existing cheats available helps with developing countermeasures, but the damage that easy availability of cheats can do to a community should not be underestimated.
Anarchy servers are a cool thing though and I'm glad you're having fun with it.

> P.S. I have tried to make an account on the MT forums like thrice and was always instantly banned not sure why lol (maybe because i use tor ?)

The forums has plugins for spam prevention, Tor is probably the worst you could choose. If you absolutely want to stay "hidden", registering via VPN should work fine.
